### PR TITLE
Fix let-syntax processing of CRLF-encoded source files

### PR DESCRIPTION
**Pull request by dra27**
In short: the existing let-syntax processor causes files containing `\r\n` to be written as `\r\r\n` (i.e. double-encoded newlines). This is a problem for the PPX driver fragments in `src/preprocessing.ml` which can then no longer be parsed.

The solution I propose here is based on the observation that the let-syntax lexer only ever emits a string with a newline sequence at the end and never manually creates a `\r` character (that's the basis of the two `assert`s). This allows easy stripping of `\r` from the string to be written so the processor now only ever writes `\n`. On Windows, `stdout` will be in text mode, so the `\n` will be translated back to `\r\n`.

The overall effect is that on Unix the resulting output will always use Unix line endings and on Windows the output will always use Windows line endings, regardless of what was in the input.

Note that naked `\r` characters (that is `\r` not followed by `\n`) was already an error and remains so.

It would also be possible to alter the Jbuild and Dune lexers to recognise newline with a more liberal `\r*\n` (which is what OCaml's lexer does), but I don't have a strong preference either way on that.

### CODE REVIEW

**Comment by emillon** [comment]
I've noticed before (and @rgrinberg too) that sometimes errors in preprocessed files were off by one line. Does this fix this issue as well?

**Comment by diml** [comment]
> On Windows, stdout will be in text mode, so the \n will be translated back to \r\n.

Since we open the input file in binary mode, shouldn't we just do `set_binary_mode_out stdout true`?

**Comment by dra27** [comment]
@diml - I'm happy to say that on this rare occasion, I did consider that simpler fix before the harder one 😄 (usually a certain Xavier points out these simpler solutions long after I've spent hours on the complicated one...). However, here it would result in mixed ending files, which are very dirty (I think it would work, but it's against my philosophy to promote generating mixed-ending files!). Having enabled binary output, you then need to ensure that any generated newlines are `\r\n`, if appropriate, which I think would be more brittle than this solution. My solution here is relatively brittle (hence the asserts), but I expect that this processor will not change much, which is why I didn't implement the longer full translation function for strings.

**Comment by dra27** [comment]
@emillon - it shouldn't be the cause, but I suppose it might. In theory, your line counting should be tracking `\n` so the extra `\r` shouldn't get in the way - but of course if `\r\r\n` has been completely ignored as a newline (because it doesn't match `\r?\n`) then I guess you could get an off-by-one in an error message?

**Comment by diml** [comment]
I see. Well if it works for you that's fine by me.

File: src/let-syntax/lexer.mll [hunk 1]
```diff
@@ -28,12 +28,21 @@ let fail lb msg =
...
-  output_string t.oc s
+  let len = String.length s in
+  if len > 0 then
+    if s.[len - 1] = '\n' then begin
+      assert (String.index s '\n' = len - 1);
+      t.line <- t.line + 1;
+      if len > 1 && s.[len - 2] = '\r' then begin
+        output_substring t.oc s 0 (len - 2);
+        output_char t.oc '\n'
+      end else
```
Comment (diml) [inline src/let-syntax/lexer.mll]: Note that this is not equivalent to the old code as we were scanning the whole string

Comment (diml) [reply src/let-syntax/lexer.mll]: Ah, indeed

**Pull request merged**